File: java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/schemas/SortDirection.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.schemas;

import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;

/**
 * Describes the sort order direction.
 */
public enum SortDirection {
    /**
     * Sorts from the lowest to the highest value.
     */
    ASCENDING(0),

    /**
     * Sorts from the highests to the lowest value.
     */
    DESCENDING(1);

    public static final int BYTEST = Integer.BYTES;

    private static Int2ObjectMap<SortDirection> mappings;
    private int value;

    SortDirection(int value) {
        this.value = value;
        mappings().put(value, this);
    }

    public int value() {
        return this.value;
    }

    public static SortDirection from(int value) {
        return mappings().get(value);
    }

    private static Int2ObjectMap<SortDirection> mappings() {
        if (mappings == null) {
            synchronized (SortDirection.class) {
                if (mappings == null) {
                    mappings = new Int2ObjectOpenHashMap<>();
                }
            }
        }
        return mappings;
    }
}

File: src/Serialization/HybridRow/Schemas/SortDirection.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#pragma warning disable CA1028 // Enum Storage should be Int32

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas
{
    using System.Runtime.Serialization;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Converters;

    /// <summary>Describes the sort order direction.</summary>
    [JsonConverter(typeof(StringEnumConverter), true)] // camelCase:true
    public enum SortDirection : byte
    {
        /// <summary>Sorts from the lowest to the highest value.</summary>
        [EnumMember(Value = "asc")]
        Ascending = 0,

        /// <summary>Sorts from the highest to the lowest value.</summary>
        [EnumMember(Value = "desc")]
        Descending,
    }
}
